### PR TITLE
This PR contains the changes necessary to enable batch release.

### PR DESCRIPTION
**Pull request by JimSuplizio**
**- sdk/template/ci.yml** - Added EnableBatchRelease: true to opt-in to batch release.
**- eng/pipelines/templates/jobs/ci.yml** - Added set-test-pipeline-version.yml for template libraries two and three. This was necessary because of the way template's version is manipulated prior to release. We need to do the same for all three template libraries.
**- eng/pipelines/templates/stages/archetype-sdk-client.yml** - The only changes here are the addition of the archetype-java-release-batch.yml if EnableBatchRelease is true. In this case we'll trim down the Artifacts list from the full list in an SDK/area's ci.yml, to only the list where releaseInBatch is true. 
**- eng/pipelines/templates/stages/archetype-java-release.yml** - The biggest change here was to fully path out /eng/pipelines/templates/variables/globals.yml instead of using relative paths. Also, the PublishDocsMs was job was missing the globals.yml template which was causing a Component Detection stage to be injected where it wasn't needed.
**- eng/pipelines/templates/stages/archetype-java-release-batch.yml** - The major changes were for restructuring. Since libraries aren't going to be released one at a time, instead of setting up a job per artifact the job can be setup and artifact processing can just be stages within the job. This removes a lot of setup and machine utilization for areas with larger numbers of libraries.
**- eng/versioning/verify_release_set.ps1** - The existing eng/pipelines/templates/steps/compilation.check.yml used a maven compile to ensure all of the dependencies for a library have been released prior to publishing. Because we're trying to release libraries, in bulk, for an SDK/area the compilation.check.yml is no longer valid. verify_release_set.ps1 takes the list of artifacts to be released and the service directory and verifies that the set of artifacts has full transitive closure. This means that the artifacts to be released won't contain an artifact whose dependency isn't slated to be released as part of the list.

### CODE REVIEW

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:29]: Should this be refactored and shared? In general one of my concerns here will be around duplication and sharing between the different release templates.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:68]: I don't expect that this will work. You can probably directly include ` length(parameters.Artifacts)` here though.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:148]: As part of this batching refactoring we should try and reduce down to one deployment stage with one environment so that we only have one approval button. I know currently it will let you approve all at once but it seems silly to have to multiple approvals for one release.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:198]: I know we talked about figuring out a way to run these in parallel but before we over optimize on that we should make the publish script handle multiple artifacts.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:317]: I assume we plan to process these in an inner loop instead of one at a time.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:359]: Same question here. I assume this will become an inner loop inside the scripts.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml:68]: @weshaggard this [does actually work](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1517042&view=results). Initially, I wasn't sure if it needed double curly braces or single square braces until I'd tested it.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (hallipr) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:111]: We're doing set-test-pipeline-version for the 2 new template packages here.  Do we need to add set-test-pipeline-version for the 2 new packages over in archetype-java-release.yml?

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (benbp) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml:68]: Interesting, not what I would have expected either.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml:111]: Good catch, I need to add those to archetype-java-release.yml as well. That got lost when I'd updated things for the "opt-in" approach. I'll get this updated in the next commit.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml:338]: Not for this particular loop. This is exactly the same setup that @hallipr had added to the nightly docs publishing. I can let him speak to why this isn't batched but I'll say that the processing in this template is less than 3 seconds per library. I don't know if this is worth any turbulence at this time.

Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml:338]: For this, yes. That loop was moved down so the PackageInfoLocations is what's being set. I was waiting to mimic the changes @hallipr had made to release yml on Tuesday. I'd needed that set of changes in before I could make mine which were done in the first commit this morning.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml]: I changed it to displayName: 'Releasing: ${{ length(parameters.Artifacts) }} libraries' which also works just fine.

File: eng/pipelines/templates/stages/archetype-sdk-client.yml
Comment (alzimmermsft) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:110]: This bit about checkboxes is no longer relevant as this PR doesn't contain that functionality

**Review by alzimmermsft — APPROVED** [review]

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:112]: It is worth considering removing this loop on steps and do the loop inside of the script or the inline script around it just to reduced the number of steps in the pipeline. Having one or more steps per library adds a lot of noise in the DevOps UI where we can do this inline in one step and just have the log for that one step output all the results that will make it easier to spot issues.

Comment (weshaggard) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml:112]: I should add this didn't use to be an issue because there was only one artifact but now when batching we will have like 40 for spring and so there will be 40 steps which will be difficult to follow and see and we will end up with a huge scrolling list in DevOPs UI. I think we should do that for most of the places in this template. We should attempt to eliminate all cases of `${{ each artifact in parameters.Artifacts }}:`

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:162]: Similarly we should eliminate this template loop and just download all the artifacts at one go and try to use the `'${{ convertToJson(parameters.Artifacts) }}'` pattern inline instead.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:285]: left over todo?

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml:362]: Even if we don't do the loop fully inside we should still move from the template loop to an inline script loop to eliminate all the noise for the repeated steps in the DevOPs job.

Comment (weshaggard) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml:362]: For this particular case the publish-blobs.yml is pretty generic and we can just download all the artifacts once and then call it once to upload everything.

File: eng/pipelines/templates/stages/archetype-sdk-client.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:112]: Is there any reason this wouldn't work for both batch and non-batch cases? That way we can eliminate the duplication and just do the if condition around which of the templates to include but everything else will match?

File: eng/versioning/verify_release_set.ps1
Comment (weshaggard) [inline eng/versioning/verify_release_set.ps1:20]: Color won't show in the devops logs but if you want this to be an error and standout we should output as an error. We have some common logging functions at https://github.com/Azure/azure-sdk-for-java/blob/main/eng/common/scripts/logging.ps1 that will toggle correctly when output in DevOps or not.

File: eng/versioning/verify_release_set.ps1
Comment (weshaggard) [inline eng/versioning/verify_release_set.ps1:27]: Any reason to not make this a hashset?

File: eng/versioning/verify_release_set.ps1
Comment (weshaggard) [inline eng/versioning/verify_release_set.ps1:62]:
```suggestion
                        $libraryToVerify = $groupId + ":" + $artifactId
```

File: eng/versioning/verify_release_set.ps1
Comment (weshaggard) [inline eng/versioning/verify_release_set.ps1:68]: LogError using the logging helepers.
```suggestion
                            LogError "$($pomFile) contains a current dependency, groupId=$($groupId), artifactId=$($artifactId), which is not the list of libraries to be released."
```

File: eng/versioning/verify_release_set.ps1
Comment (weshaggard) [inline eng/versioning/verify_release_set.ps1:38]: Generally better to prefer a foreach(...) loop over the ForEach-Object pipeline cmdlet.

File: eng/versioning/verify_release_set.ps1
Comment (weshaggard) [inline eng/versioning/verify_release_set.ps1:79]: nit: I just like `!` over `-not`
```suggestion
    if (!$script:FoundPomFile) {
```

**Review by weshaggard — CHANGES_REQUESTED** [review]
I think we still need to move more of the template loops into scripts to avoid an explosion of steps that make it very difficult to read in the DevOps UI. Even if we don't take the loop all the way to the inner script, we should use an inline script to loop over the other scripts to avoid all these extra steps.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml:285]: Yeah, that todo is done. The next commit will have this comment removed.

File: eng/versioning/verify_release_set.ps1
Comment (JimSuplizio) [reply eng/versioning/verify_release_set.ps1:62]: At least the spelling error was consistent, and it was processing correctly. This will be updated with the next commit.

File: eng/versioning/verify_release_set.ps1
Comment (JimSuplizio) [reply eng/versioning/verify_release_set.ps1:79]: Okay. This will be updated with the next commit.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml:199]: @weshaggard there's a couple of reasons why I don't want to make any changes to the publish script. Right now, I think we're better off holding off on any changes to the publish script until we know what ESRP looks like and rewriting it at that time.

File: eng/pipelines/templates/stages/archetype-sdk-client.yml
Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-sdk-client.yml:110]: Removed. It'll be updated with the next commit.

File: eng/versioning/verify_release_set.ps1
Comment (JimSuplizio) [reply eng/versioning/verify_release_set.ps1:27]: Six of one, half dozen of the other. I'll updated both libraryReleaseList and missingLibraries to be dictionaries with the next commit.

File: eng/pipelines/templates/stages/archetype-sdk-client.yml
Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-sdk-client.yml:112]: Trying this with the next commit.

File: eng/versioning/verify_release_set.ps1
Comment (JimSuplizio) [reply eng/versioning/verify_release_set.ps1:20]: Yup, moved to LogError with the next commit.

File: eng/pipelines/templates/stages/archetype-sdk-client.yml
Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-sdk-client.yml:111]: The answer to this is no. From everything I can see, the templates and parameters need to be in the same if/else block. Even tinkering around with the spacing didn't work. Having the parameters hoisted outside of the if/else for template selection made it such that none of the parameters were being passed.

**Comment by hallipr** [comment]
> we should use an inline script to loop over the other scripts to avoid all these extra steps

I think we should simplify our yaml in favor of more orchestration in powershell. I'd prefer to see that as a separate refactor PR and not expand this PR into step consolidation via powershell

File: eng/versioning/verify_release_set.ps1
Comment (JimSuplizio) [reply eng/versioning/verify_release_set.ps1]: Already committed.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml:154]: This was done in one of the last several commits. With the exception of signing, whose environment is esrp, I've changed all of the other environments to maven which results in release only having one button.
image

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml:174]: I did this for the download here. create-tags-and-git-release still requires the loop.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:209]: We shouldn't need to do the download multiple times nor the gpg setup work. We can download the entire artifact set at once.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:133]: [in follow-up] we should improve this to verify them as a batch but it will require changes to eng/common.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:175]: [in follow-up] This tagging and release creation should be updated to handle a batch.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:214]: [in follow-up] We should dig in deeper to look at cleaning up the publishing step duplication here as well as potential parallezation.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:344]: [in follow-up] We should try to avoid duplicating these steps. This looks to be all in the java repo so should be easier to fix the template.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml]: All gone.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:430]: [in follow-up] We should do this publishing as a batch.

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release-batch.yml:376]: [in follow-up] We should refactor this so it works with a set.

**Review by weshaggard — APPROVED** [review]
Changes look good. Please add follow-up issues for things tagged with [in follow-up].

File: eng/pipelines/templates/stages/archetype-java-release-batch.yml
Comment (JimSuplizio) [reply eng/pipelines/templates/stages/archetype-java-release-batch.yml:430]: This one was marked twice and is covered in the bug, https://github.com/azure/azure-sdk-for-java/issues/28662, created to deal with publish in general.

**Pull request merged**
